### PR TITLE
feat: New plugin made to support LuckPerms

### PR DESCRIPTION
**Pull request by RobotHanzo**
### Motivation
In PR #1253 CloudPerms was removed, making LuckPerms one of the best alternatives
But it lacks the feature of service / task exclusive permission 
Therefor an optional plugin was made for those who want to use LuckPerms and use task exclusive permissions

### Modification
Added a new plugin named CloudNet-LuckPerms-Addon to add some LuckPerms context (explained below)

### Result
image
Four types of context were added: `task`, `service`, `environment` and `node`
Server admins may then setup context in their permissions if needed, to make a permission service / task exclusive
(Note: I only made the plugins for platforms which both CloudNet and LuckPerms support, and this has been tested on Paper & 
Velocity)

##### Other context
N/A

### CODE REVIEW

File: plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/CloudNetContextCalculator.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import lombok.NonNull;
+import net.luckperms.api.context.ContextConsumer;
+import net.luckperms.api.context.StaticContextCalculator;
+
+public class CloudNetContextCalculator implements StaticContextCalculator {
+
+  private final ServiceInfoHolder serviceInfoHolder;
+
+  @Inject
+  public CloudNetContextCalculator(ServiceInfoHolder serviceInfoHolder) {
```
Comment (0utplay) [inline plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/CloudNetContextCalculator.java]: I think we can replace the ServiceInfoHolder with the WrapperConfiguration.

Also add `@NonNull` here

File: plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/CloudNetContextCalculator.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+package eu.cloudnetservice.plugins.luckperms;
+
+import eu.cloudnetservice.wrapper.holder.ServiceInfoHolder;
+import jakarta.inject.Inject;
+import lombok.NonNull;
+import net.luckperms.api.context.ContextConsumer;
+import net.luckperms.api.context.StaticContextCalculator;
+
+public class CloudNetContextCalculator implements StaticContextCalculator {
```
Comment (0utplay) [inline plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/CloudNetContextCalculator.java]: Add `@Singleton` here

File: plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/BungeecordLuckPermsPlugin.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  authors = "CloudNetService",
+  version = "{project.build.version}",
+  description = "Brings LuckPerms support to all server platforms",
+  dependencies = @Dependency(name = "LuckPerms")
+)
+public class BungeecordLuckPermsPlugin implements PlatformEntrypoint {
+  private final ServiceInfoHolder serviceInfoHolder;
+
+  @Inject
+  public BungeecordLuckPermsPlugin(ServiceInfoHolder serviceInfoHolder) {
```
Comment (0utplay) [inline plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/BungeecordLuckPermsPlugin.java]: Can we inject the `CloudNetContextCalculator` directly here?

Add `@NonNull` here too

**Review by 0utplay — CHANGES_REQUESTED** [review]

**Comment by RobotHanzo** [comment]
Done.

**Comment by derklaro** [comment]
Just want to throw this information here: we're planning on a seperate module to better support luckperms (for example auto downloading for the environments). I think this plugin would fit more into the module (or, we can leave it as a plugin but the LP code will be in a seperate repo anyway).

So this PR won't be merged here, but having that functionality seems like a great addition.

**Comment by RobotHanzo** [comment]
> Just want to throw this information here: we're planning on a seperate module to better support luckperms (for example auto downloading for the environments). I think this plugin would fit more into the module (or, we can leave it as a plugin but the LP code will be in a seperate repo anyway).
> 
> So this PR won't be merged here, but having that functionality seems like a great addition.

Alright, I'd be glad to move this feature into the module when it's made

**Comment by MarkusTieger** [comment]
How about making it a CloudNet-Module instead of an addon with added capabilities of managing luckperms with the cloudnet console for example?

So would love it if the Node itself also used LuckPerms for permissions, even if it's only optional.

**Comment by 0utplay** [comment]
Hi @RobotHanzo,
it took longer than we expected. But currently I am working the last preparations to remove cloudperms. The idea to create a new luckperms module is on ice (at least for now). So I would like to include the plugin you developed.

Can you maybe update this pr to RC10 and the latest minecraft version (1.20.4 is enough for now). If you don't have the time to do that, just leave a message and I will try to update the pr.

**Comment by RobotHanzo** [comment]
> Hi @RobotHanzo, it took longer than we expected. But currently I am working the last preparations to remove cloudperms. The idea to create a new luckperms module is on ice (at least for now). So I would like to include the plugin you developed.
> 
> Can you maybe update this pr to RC10 and the latest minecraft version (1.20.4 is enough for now). If you don't have the time to do that, just leave a message and I will try to update the pr.

Done

**Comment by 0utplay** [comment]
Please exec `gradlew spotlessApply` once

File: build-extensions/src/main/kotlin/Files.kt [hunk 4]
```diff
@@ -27,6 +27,7 @@ object Files {
   const val chatPlugin = "cloudnet-chat.jar"
   const val papiExpansion = "cloudnet-papi-expansion.jar"
   const val simpleNameTagsPlugin = "cloudnet-simplenametags.jar"
+  const val luckPermsAddonPlugin = "cloudnet-luckperms-addon.jar"
```
Comment (0utplay) [inline build-extensions/src/main/kotlin/Files.kt]: I think cloudnet-luckperms.jar is enough.

File: plugins/build.gradle.kts [hunk 5]
```diff
@@ -18,9 +18,12 @@ import org.apache.tools.ant.filters.ReplaceTokens
 
 subprojects {
   repositories {
-    maven("https://repo.spongepowered.org/maven/")
```
Comment (0utplay) [inline plugins/build.gradle.kts:21]: Do we really need all these repo changes here?

File: plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/BukkitLuckPermsPlugin.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import eu.cloudnetservice.ext.platforminject.api.stereotype.PlatformPlugin;
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+import lombok.NonNull;
+import net.luckperms.api.LuckPermsProvider;
+
+@Singleton
+@PlatformPlugin(
+  platform = "bukkit",
+  name = "CloudNet-LuckPerms-Addon",
```
Comment (0utplay) [inline plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/BukkitLuckPermsPlugin.java]: I think we dont need the addon suffix (this applies to all platforms)

File: plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/FabricLuckPermsPlugin.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import net.luckperms.api.LuckPermsProvider;
+
+@Singleton
+@PlatformPlugin(
+  platform = "fabric",
+  name = "CloudNet-LuckPerms-Addon",
+  version = "{project.build.version}",
+  dependencies = {
+    @Dependency(name = "fabricloader", version = ">=0.14.17"),
+    @Dependency(name = "minecraft", version = "~1.19.4"),
```
Comment (0utplay) [inline plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/FabricLuckPermsPlugin.java]: This should be updated to at least >= 1.19.4

File: plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/SpongeLuckPermsPlugin.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import lombok.NonNull;
+import net.luckperms.api.LuckPermsProvider;
+
+@Singleton
+@PlatformPlugin(
+  platform = "sponge",
+  name = "CloudNet-LuckPerms-Addon",
+  version = "{project.build.version}",
+  dependencies = {
+    @Dependency(name = "spongeapi", version = "8.0.0"),
```
Comment (0utplay) [inline plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/SpongeLuckPermsPlugin.java]: I think we are on api 9 currently (after cperms removal api 10 i think)

Comment (RobotHanzo) [reply plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/SpongeLuckPermsPlugin.java]: I thought bridge was on 8 but ok

File: plugins/build.gradle.kts (hunk 5 above)
Comment (RobotHanzo) [reply plugins/build.gradle.kts:21]: Accidental merge conflicts ig

File: plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/SpongeLuckPermsPlugin.java (hunk 8 above)
Comment (0utplay) [reply plugins/luckperms-addon/src/main/java/eu/cloudnetservice/plugins/luckperms/SpongeLuckPermsPlugin.java]: Hmm you seem to be right. I just looked at the api dependency we are adding.
I will undo that change myself.

**Review by 0utplay — DISMISSED** [review]
Looks good to me. I just have one question left. Is there a reason we don't support groups in the context? I thought luckperms would support multiple values for the same context key.

**Comment by RobotHanzo** [comment]
I just discovered that context suggestions are supported, so I added that (Which I believe adds our contexts to the web panel for easier access)

File: plugins/luckperms/src/main/java/eu/cloudnetservice/plugins/luckperms/CloudNetContextCalculator.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+public class CloudNetContextCalculator implements StaticContextCalculator {
+
+  private final WrapperConfiguration wrapperConfiguration;
+
+  @Inject
+  public CloudNetContextCalculator(@NonNull WrapperConfiguration wrapperConfiguration) {
+    this.wrapperConfiguration = wrapperConfiguration;
+  }
+
+  private ContextSet getContextSet() {
```
Comment (0utplay) [inline plugins/luckperms/src/main/java/eu/cloudnetservice/plugins/luckperms/CloudNetContextCalculator.java]:
```suggestion
  private @NonNull ContextSet contextSet() {
```

File: plugins/luckperms/src/main/java/eu/cloudnetservice/plugins/luckperms/CloudNetContextCalculator.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    contextSet.add("environment", serviceId.environmentName());
+    for (var group : this.wrapperConfiguration.serviceConfiguration().groups()) {
+      contextSet.add("group", group);
+    }
+    return contextSet;
+  }
+
+  @Override
+  public void calculate(@NonNull ContextConsumer consumer) {
+    consumer.accept(getContextSet());
```
Comment (0utplay) [inline plugins/luckperms/src/main/java/eu/cloudnetservice/plugins/luckperms/CloudNetContextCalculator.java]:
```suggestion
    consumer.accept(this.contextSet());
```

File: plugins/luckperms/src/main/java/eu/cloudnetservice/plugins/luckperms/CloudNetContextCalculator.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  @Override
+  public @NonNull ContextSet estimatePotentialContexts() {
+    return getContextSet();
```
Comment (0utplay) [inline plugins/luckperms/src/main/java/eu/cloudnetservice/plugins/luckperms/CloudNetContextCalculator.java]:
```suggestion
    return this.contextSet();
```

**Review by 0utplay — CHANGES_REQUESTED** [review]

**Review by 0utplay — APPROVED** [review]

**Comment by 0utplay** [comment]
Thanks for the contribution!

**Pull request merged**